components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/RegistryPolicyDAOImpl.java

```diff
@@ -0,0 +1,1248 @@
+        }
+
+        if (policyDTOs != null) {
+            attributeMap = new HashMap<>();
+            for (PolicyDTO policyDTO : policyDTOs) {
+                Set<AttributeDTO> attributeDTOs =
+                        new HashSet<>(Arrays.asList(policyDTO.getAttributeDTOs()));
+                String[] policyIdRef = policyDTO.getPolicyIdReferences();
+                String[] policySetIdRef = policyDTO.getPolicySetIdReferences();
+
+                if (policyIdRef != null && policyIdRef.length > 0 || policySetIdRef != null &&
+                        policySetIdRef.length > 0) {
+                    for (PolicyDTO dto : policyDTOs) {
+                        if (policyIdRef != null) {
+                            for (String policyId : policyIdRef) {
+                                if (dto.getPolicyId().equals(policyId)) {
+                                    attributeDTOs.addAll(Arrays.asList(dto.getAttributeDTOs()));
+                                }
+                            }
+                        }
+                        for (String policySetId : policySetIdRef) {
+                            if (dto.getPolicyId().equals(policySetId)) {
+                                attributeDTOs.addAll(Arrays.asList(dto.getAttributeDTOs()));
+                            }
+                        }
+                    }
+                }
+                attributeMap.put(policyDTO.getPolicyId(), attributeDTOs);
+            }
+        }
+
+        return attributeMap;
+    }
+
+    /**
+     * Gets support attribute searching scheme of the module.
+     *
+     * @return return scheme identifier value
+     */
+    @Override
+    public int getSupportedSearchAttributesScheme() {
+
+        return PolicyFinderModule.COMBINATIONS_BY_CATEGORY_AND_PARAMETER;
+    }
+
+    /**
+     * Lists all PAP policy IDs.
+     *
+     * @return list of policy IDs
+     * @throws EntitlementException If an error occurs
+     */
+    @Override
+    public List<String> listPolicyIds() throws EntitlementException {
+
+        String path = PDPConstants.ENTITLEMENT_POLICY_PAP;
+        return listAllPolicyIds(path);
+
+    }
+
+    /**
+     * Removes the given policy from PAP.
+     *
+     * @param policyId policy ID
+     * @throws EntitlementException If an error occurs
+     */
+    @Override
+    public void removePolicy(String policyId) throws EntitlementException {
+
+        String path;
+
+        if (log.isDebugEnabled()) {
+            log.debug("Removing entitlement policy");
+        }
+
+        // Restricts removing policies, that have already been published
+        List<String> publishedPolicies = listPublishedPolicyIds();
+        if (publishedPolicies != null && publishedPolicies.contains(policyId)) {
+            log.error("Policies that have already been published, cannot be removed from PAP");
+            throw new EntitlementException("Policies that have already been published, cannot be removed from PAP");
+        }
```

Comment:
Check if this is an informed decision. This is a change from old behavior where it was allowed to delete already published policies from PAP.
Integration test `OAuth2XACMLScopeValidatorTestCase` fails due to this.

Reply:
This change is to cater to the db impl where separate removal of a policy from PDP and PAP is not possible. So will change the integration test with https://github.com/wso2/product-is/pull/20512